src/main/kotlin/biokotlin/kmer/Kmer.kt

```diff
@@ -0,0 +1,143 @@
+    var encoding = 0L
+    seq.forEach{ c -> encoding = (encoding shl 2) or (c.code.toLong() ushr 1 and 3L) }
+    return Kmer(encoding)
+}
+
+/**
+ * Create a Kmer from a NucSeq
+ * Sequence may be fewer than 32 bases long, left will be padded with A's to equal 32 bases
+ * @param seq NucSeq without ambiguous bases
+ */
+fun Kmer(seq: NucSeq): Kmer {
+    return Kmer(seq.seq())
+}
+
+/**
+ * 2-bit encoded representation of a short nucleotide sequence (32 bases or fewer)
+ * Encoding: A=0, C=1, T/U=2, G=3
+ * For kmers smaller than 32 bp, leftmost digits are padded with 0
+ * Ambiguous bases (eg. N) are not allowed
+ *
+ *
+ * ``` kotlin
+ * val kmer = Kmer(1432462)
+ * val kmerFromString = Kmer("UGCACUGCCA")
+ * val kmerFromSeq = Kmer(NucSeq("TCTCACCTACA")
+ *
+ * val kmerFromString = Kmer("NTACCGANNTCG") // throw error - ambiguous bases
+ * val kmerFromString = Kmer("TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTT") // throw error - string too long
+ * ```
+ */
+@JvmInline
+value class Kmer(val encoding: Long): Comparable<Kmer> {
+
+    /** Follows comparisons of bits represented as unsigned long */
+    override operator fun compareTo(other: Kmer): Int {
+        return encoding.toULong().compareTo(other.encoding.toULong())
+    }
+
+    /**
+     * Converts 2-bit encoding to string
+     * By default, keeps any A padding at left
+     */
+    override fun toString(): String {
+        return toString(32)
+    }
+
+    /**
+     * Converts 2-bit encoding to string, trimming leftmost bits
+     * so the resulting sequence length is [kmerSize]
+     */
+    fun toString(kmerSize: Int): String {
+        var temp = encoding
+        return (0 until kmerSize).map{
+            val y = when ( temp and 3L) {
+                0L -> 'A'
+                1L -> 'C'
+                2L -> 'T'
+                3L -> 'G'
+                else -> 'N'
+            }
+            temp = temp ushr 2
+            return@map y
+        }.reversed().joinToString("")
+    }
+
+    /**
+     * Converts Kmer to NucSeq
+     * Includes any A padding at left
+     */
+    fun toSeq(): NucSeq {
+        return Seq(this.toString())
+    }
+
+    /**
+     * Converts Kmer to NucSeq, trimming A padding at left to [kmerSize]
+     */
+    fun toSeq(kmerSize: Int): NucSeq {
+        return Seq(this.toString(kmerSize))
+    }
+
+    /**
+     * Returns the reverse complement of a sequence already encoded in a 2-bit long.
+     * Note: polyA is used represent unknown, but reverse complement will change it to polyT which does not mean the same.
+     * [kmerSize] parameter is required to remove polyT padding
+     */
+    fun reverseComplement(kmerSize: Int): Kmer {
+
+        // reverse bits
+        // no need to reverse the two bits encoding the same nucleotide
+        // eg. TAGC -> CGAT
+        var y = ((encoding and 0x3333333333333333) shl 2) or ((encoding and -3689348814741910324) ushr 2)
```

Comment:
Change the negative magic numbers to also be hex?

Reply:
I could, though the hex would also have to be negative. It is somewhat more readable like that though.